### PR TITLE
Deprecate subset getters

### LINKED ISSUE #2224: Deprecate certain "get_..." methods
**Reporter: Jesse Averbukh**

The following methods need to be deprecated and all instances switched with their respective replacements:
 * get_subsets_from_viewer() replaced with get_subsets()
 * get_data_from_viewer() replaced with get_data()
 * specviz.get_spectra(subset_to_apply) replaced with specviz.get_spectra(spectral_subset)

 

Updating the docs can be part of this ticket or a separate effort.

 --- 

 DISCLAIMER: This issue was autocreated by the Jdaviz Issue Creation Bot on behalf of the reporter. If any information is incorrect, please contact [Duy Nguyen](https://github.com/duytnguyendtn)

### PR DESCRIPTION
**Pull request by duytnguyendtn**
### Description

This PR identifies existing locations in our codebase that utilize our old subset and data getters and replaces it with our new get_subsets/getdata methods. Specifically, the following three changes:

1. get_subsets_from_viewer() replaced with get_subsets()
2. get_data_from_viewer() replaced with get_data()
3. specviz.get_spectra(subset_to_apply) replaced with specviz.get_spectra(spectral_subset)

In addition, I had to fix a bug with the `Mosviz2DProfileViewer.get_data()` method where it was returning `statistic=mean` of the Spectrum1D object by default due to not having the ability to specify the `statistic`. I added a keyword argument and changed the default to `statistic=None`

I think I got all the docs changes, but let me know if I missed something!

FYI decreased coverage is due to the `get_subsets_from_viewer` and `get_data_from_viewer` methods not being covered by tests anymore

Fixes #2224

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/app.py [hunk 1]
```diff
@@ -753,6 +754,9 @@ def get_subsets_from_viewer(self, viewer_reference, data_label=None, subset_type
             representing the subset name and values as astropy regions
             objects.
         """
+        warnings.warn(AstropyDeprecationWarning("get_subsets_from_viewer() is deprecated in v3.6 "
```
[the remark this answers is not in the document]
Comment (duytnguyendtn) [reply jdaviz/app.py:757]: Ahh, didn't know that was a thing. Thanks!

**Comment by duytnguyendtn** [comment]
> Also, are you feeling adventurous? 😆
> 
> [#2201 (review)](https://github.com/spacetelescope/jdaviz/pull/2201#pullrequestreview-1468300659)
> 
> @javerbukh , "... do we want to keep using `get_interactive_regions()` over `get_subsets()`?"

Falling a little behind schedule so I'm going to leave this work out

**Comment by javerbukh** [comment]
Not a blocker but when I do `specviz.app.get_subsets_from_viewer("spectrum-viewer")` I see
```
WARNING: AstropyDeprecationWarning: The get_subsets_from_viewer function is deprecated and may be removed in a future version.
        Use get_subsets instead. [warnings]
WARNING:astroquery:AstropyDeprecationWarning: The get_subsets_from_viewer function is deprecated and may be removed in a future version.
        Use get_subsets instead.
WARNING: AstropyDeprecationWarning: The get_data_from_viewer function is deprecated and may be removed in a future version.
        Use get_data instead. [jdaviz.app]
WARNING:astroquery:AstropyDeprecationWarning: The get_data_from_viewer function is deprecated and may be removed in a future version.
        Use get_data instead.
```
I understand why the warning was issued for `get_data_from_viewer` as well but why were both warnings printed twice?

**Comment by pllim** [comment]
`astroquery:AstropyDeprecationWarning` ??? Why is it calling `astroquery`?

**Comment by duytnguyendtn** [comment]
I'm not sure; the even weirder thing is that it's warning is coming from `astroquery`??

**Comment by duytnguyendtn** [comment]
`app.py` doesn't even import anything from astroquery!!

**Comment by pllim** [comment]
Try turning warning into error in your session and inspect the traceback. Maybe somewhere deep in the machinery, one deprecated function is calling another deprecated function.

File: docs/cubeviz/export_data.rst [hunk 2]
```diff
@@ -41,7 +41,7 @@ To get all subsets from the spectrum viewer:
 
 .. code-block:: python
 
-    subset1_spec1d = cubeviz.app.get_subsets_from_viewer("spectrum-viewer")
+    subset1_spec1d = cubeviz.app.get_subsets()
```
Comment (javerbukh) [inline docs/cubeviz/export_data.rst:44]:
```suggestion
    subset1_spec1d = cubeviz.specviz.app.get_subsets()
```
This will return all the subsets that specviz handles, which is the same as all the subsets from the spectrum viewer. You can maybe reword the above as "all spectral subsets" instead of "from the spectrum viewer", although both phrases are technically true.

File: jdaviz/configs/imviz/plugins/coords_info/coords_info.py [hunk 3]
```diff
@@ -469,7 +469,7 @@ def _copy_axes_to_spectral():
                     sp = self.app._get_object_cache[cache_key]
                 else:
                     sp = self._specviz_helper.get_data(data_label=data_label,
-                                                       subset_to_apply=subset_label)
+                                                       spectral_subset=subset_label)
```
Comment (javerbukh) [inline jdaviz/configs/imviz/plugins/coords_info/coords_info.py:472]: Are you sure this is trying to get the spectral subset and not a collapsed spatial subset?

File: jdaviz/tests/test_app.py [hunk 4]
```diff
@@ -1,142 +1,143 @@
...
-    assert dc[2].label == f"{data_label} (2)"
-
-
-def test_case_that_used_to_break_return_label(specviz_helper, spectrum1d):
-    specviz_helper.load_spectrum(spectrum1d, data_label="this used to break (1)")
-    specviz_helper.load_spectrum(spectrum1d, data_label="this used to break")
-    dc = specviz_helper.app.data_collection
-    assert dc[0].label == "this used to break (1)"
-    assert dc[1].label == "this used to break (2)"
+import pytest
```
Comment (javerbukh) [inline jdaviz/tests/test_app.py:1]: I'm having a hard time figuring out what changed in this file, can you point me to the actual diff?

File: jdaviz/tests/test_subsets.py [hunk 5]
```diff
@@ -24,8 +24,8 @@ def test_region_from_subset_2d(cubeviz_helper):
 
     cubeviz_helper.app.get_viewer('flux-viewer').apply_roi(EllipticalROI(1, 3.5, 1.2, 3.3))
 
-    subsets = cubeviz_helper.app.get_subsets_from_viewer('flux-viewer')
-    reg = subsets.get('Subset 1')
+    subsets = cubeviz_helper.app.get_subsets()
+    reg = subsets.get('Subset 1')[0]['region']
```
Comment (javerbukh) [inline jdaviz/tests/test_subsets.py:28]: This is good but for posterity there is another way to get this same return:
`cubeviz_helper.app.get_subsets('Subset 1', object_only=True)[0]`
This makes it so that only the region objects are returned for spatial subsets.

File: jdaviz/tests/test_app.py (hunk 4 above)
Comment (duytnguyendtn) [reply jdaviz/tests/test_app.py:1]: Ugh... this happened again. I don't know why GitHub does this sometimes. Let me see if I can track it down (git diff on my machine is showing the same thing here)

Comment (duytnguyendtn) [reply jdaviz/tests/test_app.py:1]: @javerbukh This is the only thing that changed in that file: https://github.com/spacetelescope/jdaviz/pull/2242/commits/92c0bd8f7d99a8cf7e2de38ec3f2337d09b0468b

You can see the diff here properly! I don't know why GitHub is showing the entire file as changed...

File: jdaviz/configs/imviz/plugins/coords_info/coords_info.py (hunk 3 above)
Comment (duytnguyendtn) [reply jdaviz/configs/imviz/plugins/coords_info/coords_info.py:472]: Regardless, the arg name has changed, so that necessitates the change here. Unless this wasn't considered when the decision was made to rename the arg and we need to rethink this?

File: docs/cubeviz/export_data.rst (hunk 2 above)
Comment (duytnguyendtn) [reply docs/cubeviz/export_data.rst:44]: Ahh got catch. I was supposed to add `spectral_only=True` but in this context, I think your suggestion makes sense

File: jdaviz/tests/test_subsets.py (hunk 5 above)
Comment (duytnguyendtn) [reply jdaviz/tests/test_subsets.py:28]: Definitely good to know! Since these tests are explicitly extracting the region, I think it's useful to see where the regions are coming from in the full dictionary, but this is still great to keep in mind for more general audiences

File: jdaviz/tests/test_app.py (hunk 4 above)
Comment (kecnry) [reply jdaviz/tests/test_app.py:1]: It looks like your editor is either changing the newline or stripping out trailing whitespace on save.  Can you stage and commit only the individual lines instead of all changes?  (I'm guessing the individual commit diff is set to ignore whitespace changes).

File: jdaviz/configs/imviz/plugins/coords_info/coords_info.py (hunk 3 above)
Comment (pllim) [reply jdaviz/configs/imviz/plugins/coords_info/coords_info.py:472]: @kecnry did this part so he would know best.

File: jdaviz/configs/mosviz/plugins/viewers.py [hunk 6]
```diff
@@ -88,8 +88,9 @@ def __init__(self, *args, **kwargs):
             "spectrum_2d_viewer_reference_name", "spectrum-2d-viewer"
         )
 
-    def data(self, cls=None):
-        return [layer_state.layer.get_object(cls=cls or self.default_class)
+    def data(self, cls=None, statistic=None):
+        return [layer_state.layer.get_object(statistic=statistic,
```
Comment (pllim) [inline jdaviz/configs/mosviz/plugins/viewers.py:92]: Is this an unrelated bug fix? If so, the change log also need to reflect this fix.

Actually I don't see any change log modification here, so that has to be fixed too.

File: jdaviz/configs/mosviz/tests/test_data_loading.py [hunk 7]
```diff
@@ -42,13 +44,15 @@ def test_load_image(mosviz_helper, mos_image):
...
 
-    data = mosviz_helper.app.get_data_from_viewer('image-viewer')
+    data = mosviz_helper.app.get_viewer(mosviz_helper._default_image_viewer_reference_name
+                                        ).data()
 
-    dataval = data[f"{label} 0"]
-    assert isinstance(dataval, CCDData)
+    assert len(data) == 1
+    dataval = data[0]
+    assert isinstance(dataval, Data)
```
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_data_loading.py:55]: Wait... is this PR supposed to modify the data type that gets returned? This looks like a behavior change.

File: jdaviz/tests/test_subsets.py (hunk 5 above)
Comment (pllim) [reply jdaviz/tests/test_subsets.py:28]: Why not just change it now?

Comment (duytnguyendtn) [reply jdaviz/tests/test_subsets.py:28]: > I think it's useful to see where the regions are coming from in the full dictionary

But when we advertise to our users, I think the other syntax is better

**Review by pllim — COMMENTED** [review]
So, did that problem with multiple deprecating warnings ever got resolved? What was the problem?

File: jdaviz/configs/imviz/plugins/coords_info/coords_info.py (hunk 3 above)
Comment (kecnry) [reply jdaviz/configs/imviz/plugins/coords_info/coords_info.py:472]: yes, I believe we want spatial here (you can test it by seeing if mouseover on the live-collapsed spectrum still works)

Comment (duytnguyendtn) [reply jdaviz/configs/imviz/plugins/coords_info/coords_info.py]: Done! Fixed in https://github.com/spacetelescope/jdaviz/pull/2242/commits/5a9193e2388c12638e12e75b2f93b7a1d2e4ee92 and verified the mouseover plugin works in imviz and cubeviz. Thanks!

File: jdaviz/tests/test_app.py (hunk 4 above)
Comment (duytnguyendtn) [reply jdaviz/tests/test_app.py:1]: I don't think this is the case. Whenever I commit, there's an explicit line that converts my MSDOS line endings to UNIX line endings:
`dos2unix: converting file E:/STScI/gitRepos/jdaviz/.git/COMMIT_EDITMSG to Unix format...`

Because WYSI NOT WYG on my end, I can't really inspect what's getting committed. Ironically, I think the way to check for that is someone on the mac side open the file and see if you have a CR or LF character at the end of the line (But that being said, if it is a line character thing, I wouldn't expect the OSX or Ubuntu tests to successfully run at all

File: jdaviz/configs/mosviz/tests/test_data_loading.py (hunk 7 above)
Comment (duytnguyendtn) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:55]: The intention I read from this test is to check if the data was loaded into the viewer properly, less so the actual class. That why I couldn't swap this out with `get_data` because there's no way to filter on which viewer the data are in.

For that reason, I had to do something a little different (call `viewer.data()` instead). I don't think it was guaranteed this would result in the same class, but it verifies the actual purpose of this test

**Comment by duytnguyendtn** [comment]
> So, did that problem with multiple deprecating warnings ever got resolved? What was the problem?

Nope, though @javerbukh indicated this wasn't a blocker, so I think we're safe to proceed

**Comment by duytnguyendtn** [comment]
I think I addressed all the comments. Ready for rereview!

**Comment by pllim** [comment]
Please add a change log. Thanks!

**Comment by pllim** [comment]
Also need a rebase because that one-liner bug fix has been merged as #2258 .

File: jdaviz/configs/mosviz/tests/test_data_loading.py (hunk 7 above)
Comment (pllim) [reply jdaviz/configs/mosviz/tests/test_data_loading.py:55]: I found a way to keep the old behavior. Will push a commit.

File: jdaviz/tests/test_app.py (hunk 4 above)
Comment (pllim) [reply jdaviz/tests/test_app.py:1]: Probably the endlines. I think we have mixed endlines in the code base. I ended up blinking the copy from this branch with the copy from `main` and this is the only difference:

```python
    assert not len(viz.app.get_data_from_viewer("h", "non-existent label"))
```

```python
    with pytest.raises(ValueError):
        viz.get_data("non-existent label")
```

I think the behavior change is acceptable. The new behavior is more explicit.

**Review by pllim — APPROVED** [review]
I applied my review as commits, so LGTM now. @javerbukh will do the final review.

File: jdaviz/tests/test_app.py (hunk 4 above)
Comment (javerbukh) [reply jdaviz/tests/test_app.py:1]: No I mean this PR touching all of `test_app.py`. I'd prefer to maintain the correct git history if possible.

Comment (pllim) [reply jdaviz/tests/test_app.py:1]: Errr... I can try but I cannot promise anything.

**Comment by duytnguyendtn** [comment]
Thank you everyone for merging this in my absence!
